File: src/main/scala/net/machinemuse/numina/network/MusePacketHandler.scala
/**
 *
 */
package net.machinemuse.numina.network

import java.io._
import java.util

import io.netty.buffer.ByteBufInputStream
import io.netty.channel.ChannelHandler.Sharable
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToMessageCodec
import net.machinemuse.numina.general.MuseLogger
import net.machinemuse.numina.scala.MuseNumericRegistry
import net.minecraft.client.Minecraft
import net.minecraft.client.network.NetHandlerPlayClient
import net.minecraft.network.NetHandlerPlayServer
import net.minecraftforge.fml.common.network.NetworkRegistry
import net.minecraftforge.fml.common.network.internal.FMLProxyPacket

/**
 * @author MachineMuse
 */
@Sharable
object MusePacketHandler extends MessageToMessageCodec[FMLProxyPacket, MusePacket]{
  val networkChannelName: String = "Numina"
  val packagers = new MuseNumericRegistry[MusePackager]
  val channels = NetworkRegistry.INSTANCE.newChannel(networkChannelName, this)


  override def decode(ctx: ChannelHandlerContext, msg: FMLProxyPacket, out: util.List[AnyRef]): Unit = {
    val data = new DataInputStream(new ByteBufInputStream(msg.payload))
    var packetType: Int = 0
    try {
      msg.handler match {
        case h: NetHandlerPlayServer =>
          val player = h.playerEntity
          packetType = data.readInt
          for {// For comprehension for an option
            packager <- MusePacketHandler.packagers.get(packetType)
            packet = packager.read(data, player)
          } {
            packet handleServer player
          }
        case h: NetHandlerPlayClient =>
          val player = Minecraft.getMinecraft.thePlayer
          packetType = data.readInt
          for {// For comprehension for an option
            packager <- MusePacketHandler.packagers.get(packetType)
            packet = packager.read(data, player)
          } {
            packet handleClient player
          }
      }
    } catch {
      case e: IOException => MuseLogger.logException("PROBLEM READING PACKET IN DECODE STEP D:", e)
    }
  }

  override def encode(ctx: ChannelHandlerContext, msg: MusePacket, out: util.List[AnyRef]) = {
    out.add(msg.getFMLProxyPacket)
  }

}

File: src/main/java/net/machinemuse/numina/network/MusePacketHandler.java
package net.machinemuse.numina.network;

import io.netty.buffer.ByteBufInputStream;
import io.netty.channel.ChannelHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToMessageCodec;
import net.machinemuse.numina.general.MuseLogger;
import net.machinemuse.numina.scala.MuseNumericRegistry;
import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.EntityPlayerSP;
import net.minecraft.client.network.NetHandlerPlayClient;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.network.INetHandler;
import net.minecraft.network.NetHandlerPlayServer;
import net.minecraftforge.fml.common.network.FMLEmbeddedChannel;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.minecraftforge.fml.common.network.internal.FMLProxyPacket;
import net.minecraftforge.fml.relauncher.Side;

import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.EnumMap;
import java.util.List;

/**
 * @author MachineMuse
 *
 * Ported to Java by lehjr on 10/23/16.
 */
@ChannelHandler.Sharable
public final class MusePacketHandler extends MessageToMessageCodec<FMLProxyPacket, MusePacket>
{
    public static String networkChannelName;
    public static MuseNumericRegistry<MusePackager> packagers;
    public static EnumMap<Side, FMLEmbeddedChannel> channels;


    private MusePacketHandler() {
        this.networkChannelName = "Numina";
        this.packagers = new MuseNumericRegistry<>();
        this.channels = NetworkRegistry.INSTANCE.newChannel(this.networkChannelName, new ChannelHandler[] { this });
    }

    static {
        new MusePacketHandler();
    }

    public void encode(final ChannelHandlerContext ctx, final MusePacket msg, final List<Object> out) {
        try {
            out.add(msg.getFMLProxyPacket());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void decode(final ChannelHandlerContext ctx, final FMLProxyPacket msg, final List<Object> out) {
        final DataInputStream data = new DataInputStream((InputStream)new ByteBufInputStream(msg.payload()));
        int packetType = 0;
        try {
            final INetHandler handler = msg.handler();
            if (handler instanceof NetHandlerPlayServer) {
                final EntityPlayerMP playerServer = ((NetHandlerPlayServer)handler).playerEntity;
                packetType = data.readInt();
                MusePackager packagerServer = this.packagers.get(packetType);
                MusePacket packetServer = packagerServer.read(data, playerServer);
                packetServer.handleServer(playerServer);
            }
            else {
                if (!(handler instanceof NetHandlerPlayClient)) {
                    throw new IOException("Error with (INetHandler) handler. Should be instance of NetHandlerPlayClient.");
                }
                final EntityPlayerSP playerClient = Minecraft.getMinecraft().thePlayer;
                packetType = data.readInt();
                MusePackager packagerClient = this.packagers.get(packetType);
                MusePacket packetClient = packagerClient.read(data, playerClient);
                packetClient.handleClient(playerClient);
            }
        }
        catch (IOException exception) {
            MuseLogger.logException("PROBLEM READING PACKET IN DECODE STEP D:", exception);
        }
    }
}